### PR TITLE
Add print_stats_for_non_join_query session property

### PR DESCRIPTION
**Pull request by picrinite**
Provide an optoin to only show stats for exaplain query or
join query in query plan.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 1]
```diff
@@ -109,6 +109,7 @@
     private boolean iterativeOptimizerEnabled = true;
     private boolean enableStatsCalculator = true;
     private boolean ignoreStatsCalculatorFailures = true;
+    private boolean printStatsForNonJoinQuery = true;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: I would keep it `false` by default. Deployments with QueryEventListener that record all query plans are rather unusual, and in addition to that it is very unlikely that someone would like to see the stats for the queries that cannot be optimized by the CBO in the event log.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 2]
```diff
@@ -159,12 +159,12 @@ public LogicalPlanner(Session session,
         this.warningCollector = requireNonNull(warningCollector, "warningCollector is null");
     }
 
-    public Plan plan(Analysis analysis)
+    public Plan plan(Analysis analysis, Boolean isRawQueryExplain)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: I would pass this parameter in the constructor and call it simply `explain`

Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: Also prefer primitive `boolean`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java [hunk 3]
```diff
@@ -106,7 +106,7 @@ public Analysis analyze(Session session, Statement statement, List<Expression> p
         return analyzer.analyze(statement);
     }
 
-    public String getPlan(Session session, Statement statement, Type planType, List<Expression> parameters, WarningCollector warningCollector)
+    public String getPlan(Session session, Boolean isRawQueryExplain, Statement statement, Type planType, List<Expression> parameters, WarningCollector warningCollector)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java]: Don't add the parameter to this method signature.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java [hunk 4]
```diff
@@ -160,14 +160,14 @@ public String getJsonPlan(Session session, Statement statement, Type planType, L
...
-                Plan plan = getLogicalPlan(session, statement, parameters, warningCollector);
+                Plan plan = getLogicalPlan(session, isRawQueryExplain, statement, parameters, warningCollector);
                 return textIOPlan(plan.getRoot(), metadata, session);
             default:
                 throw new PrestoException(NOT_SUPPORTED, format("Unsupported explain plan type %s for JSON format", planType));
         }
     }
 
-    public Plan getLogicalPlan(Session session, Statement statement, List<Expression> parameters, WarningCollector warningCollector)
+    public Plan getLogicalPlan(Session session, Boolean isRawQueryExplain, Statement statement, List<Expression> parameters, WarningCollector warningCollector)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java]: This doesn't have to be parameterized. Just create `LogicalPlanner` with `explain` always set to `true`.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java [hunk 5]
```diff
@@ -176,12 +176,12 @@ public Plan getLogicalPlan(Session session, Statement statement, List<Expression
 
         // plan statement
         LogicalPlanner logicalPlanner = new LogicalPlanner(session, planOptimizers, idAllocator, metadata, sqlParser, statsCalculator, costCalculator, warningCollector);
-        return logicalPlanner.plan(analysis);
+        return logicalPlanner.plan(analysis, isRawQueryExplain);
     }
 
-    private SubPlan getDistributedPlan(Session session, Statement statement, List<Expression> parameters, WarningCollector warningCollector)
+    private SubPlan getDistributedPlan(Session session, Boolean isRawQueryExplain, Statement statement, List<Expression> parameters, WarningCollector warningCollector)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/QueryExplainer.java]: Ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 6]
```diff
@@ -125,7 +125,8 @@
     private final CostCalculator costCalculator;
     private final WarningCollector warningCollector;
 
-    public LogicalPlanner(Session session,
+    public LogicalPlanner(boolean explain,
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: nit: Put it on the next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 7]
```diff
@@ -134,10 +135,11 @@ public LogicalPlanner(Session session,
             CostCalculator costCalculator,
             WarningCollector warningCollector)
     {
-        this(session, planOptimizers, DISTRIBUTED_PLAN_SANITY_CHECKER, idAllocator, metadata, sqlParser, statsCalculator, costCalculator, warningCollector);
+        this(explain, session, planOptimizers, DISTRIBUTED_PLAN_SANITY_CHECKER, idAllocator, metadata, sqlParser, statsCalculator, costCalculator, warningCollector);
     }
 
-    public LogicalPlanner(Session session,
+    public LogicalPlanner(boolean explain,
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: ditto

File: presto-tpcds/src/test/java/com/facebook/presto/tpcds/statistics/TestTpcdsLocalStats.java [hunk 8]
```diff
@@ -40,6 +41,7 @@ public void setUp()
         Session defaultSession = testSessionBuilder()
                 .setCatalog("tpcds")
                 .setSchema("sf1")
+                .setSystemProperty(PRINT_STATS_FOR_NON_JOIN_QUERY, "true")
```
Comment (arhimondr) [inline presto-tpcds/src/test/java/com/facebook/presto/tpcds/statistics/TestTpcdsLocalStats.java]: I guess you wanted to put that in the previous commit

Comment (picrinite) [reply presto-tpcds/src/test/java/com/facebook/presto/tpcds/statistics/TestTpcdsLocalStats.java]: Good  catch.  This comes with the second commit b/c it is associated boolean explain parameter, caller is  
  LogicalQueryRunner.java#848:  LogicalPlanner logicalPlanner = new LogicalPlanner(wrappedStatement instanceof Explain....

**Comment by kokosing** [comment]
What is the motivation behind this change?

**Comment by rschlussel** [comment]
@kokosing We have some very low latency queries that stats fetching takes a non-trivial amount of the time.  We don't want to take the hit of fetching stats for queries that won't benefit from optimizations since the only stats based rules right now are joins.

**Pull request merged**